packages/nextjs/src/server/wrapServerComponentWithSentry.ts

```diff
@@ -0,0 +1,67 @@
+import { captureException, getCurrentHub, startTransaction } from '@sentry/core';
+import * as domain from 'domain';
+
+import type { ServerComponentContext } from '../common/types';
+
+/**
+ * Wraps an `app` directory server component with Sentry error instrumentation.
+ */
+export function wrapServerComponentWithSentry<F extends (...args: any[]) => any>(
+  appDirComponent: F,
+  context: ServerComponentContext,
+): F {
+  const { componentRoute, componentType } = context;
+
+  // Even though users may define server components as async functions, for the client bundles
+  // Next.js will turn them into synchronous functions and it will transform any `await`s into instances of the `use`
+  // hook. 🤯
+  return new Proxy(appDirComponent, {
+    apply: (originalFunction, thisArg, args) => {
+      return domain.create().bind(() => {
+        let maybePromiseResult;
+
+        const transaction = startTransaction({
+          op: 'function.nextjs',
+          name: `${componentType} Server Component (${componentRoute})`,
```

Comment:
I wonder if this is the name we should choose for this 🤔

Maybe we try to have the `componentRoute` at the start?

Reply:
I tried to pick something that makes it immediately obvious what the transaction is measuring. Right now it is already super confusing to see API route transactions like `GET /api/my/route` intertwined with pageload transactions like `/my/route`.

IMO having the transacitons show up as `Page Server Component (/user/[id])` and `Layout Server Component (/user/[id])` is very clear.

Reply:
Yeah that's fair, makes sense to me.